### PR TITLE
add no_break_image break mode: no partial image at the beginning of a…

### PR DESCRIPTION
**Pull request by violet-zct**
…n examples

**Patch Description**
Describe your changes

**Testing steps**
Describe how you tested your changes

### CODE REVIEW

File: metaseq/data/document_to_sequence.py [hunk 1]
```diff
@@ -193,9 +194,9 @@ def __init__(
 
         self.indices = None
 
-        # A map from document id -> number of tokens in the document.
-        # This is an atomic array because it shared among the dataloader workers and the training
-        # process.
+        # # A map from document id -> number of tokens in the document.
```
Comment (ArmenAg) [inline metaseq/data/document_to_sequence.py]: Are we adding double comments on purpose?

File: metaseq/data/document_to_sequence.py [hunk 2]
```diff
@@ -289,24 +296,26 @@ def documents() -> Iterable[Tuple[int, Document]]:
             """
             for idx in indices:
                 ln = self.len_cache.data[idx]
+                # this is a hack, it's hard to create a buffer for the irregular type of image_spans, so we query anyway
+                r, image_spans = self.dataset[idx]  
                 if ln == 0:
                     # Cache miss: we don't know the number of tokens
                     # so we have to load and tokenize the document.
-                    r = self.dataset[idx]
+                    # r, image_spans = self.dataset[idx]
```
Comment (ArmenAg) [inline metaseq/data/document_to_sequence.py]: why is r commented out?

File: metaseq/data/document_to_sequence.py [hunk 3]
```diff
@@ -524,3 +533,49 @@ def yield_token_blocks(iterable, block_size, drop_last) -> Iterable[Sequence]:
...
+            if cur_block_remain == 0:
+                yield {
+                    "ids": cur_block_ids,
+                    "block": cur_block,
+                }
+                cur_block = []
+                cur_block_ids = []
+                cur_block_remain = block_size
+
+                # if torch.distributed.get_rank() == 0:
```
Comment (ArmenAg) [inline metaseq/data/document_to_sequence.py]: let's clean this up

File: metaseq/tasks/streaming_language_modeling.py [hunk 4]
```diff
@@ -373,24 +380,31 @@ def _tokenize_interleaving_json(self, json):
...
             ra_indexes.append(ra_index)
+            image_span = (1 + cur_len + image_span[0], 1 + cur_len + image_span[1])
+            cur_len += len(ra_index)
+            image_spans.append(image_span)
+
         final_indexes = torch.cat(
             [torch.LongTensor([self.eod])] + ra_indexes + [query_index]
         )
-        return final_indexes
+        image_spans.append((1 + cur_len + query_image_span[0], 1 + cur_len + query_image_span[1]))
```
Comment (ArmenAg) [inline metaseq/tasks/streaming_language_modeling.py]: Could we add a comment why we're adding 1 (just show we don't have magical numbers)

**Comment by ArmenAg** [comment]
@violet-zct Could we add some tests that cover the behavior we expect? Similar to https://github.com/facebookresearch/metaseq/blob/main/cpu_tests/test_streaming_token_block_dataset.py

**Comment by violet-zct** [comment]
1. I pushed a cpu test for no_image_break: https://github.com/facebookresearch/metaseq/blob/cm3v2/metaseq/data/document_to_sequence.py#L295
During my tests, I fix two corner cases in get_document_boundaries() of the cm3 dataset: https://github.com/facebookresearch/metaseq/blob/cm3v2_chunting_dev/metaseq/data/cm3_dataset.py#L208 and #L211, we probably can hit these corner cases when the document is very long.

2. In addition, I found that your get_document_boundaries() can't pass the test: https://github.com/facebookresearch/metaseq/blob/cm3v2_chunting_dev/cpu_tests/test_cm3_dataset.py#L97, unless you remove + 1 in this line: https://github.com/facebookresearch/metaseq/blob/cm3v2_chunting_dev/cpu_tests/test_cm3_dataset.py#L97

3. I fixed image spans by also including the special token at the end of a image.

**Comment by ArmenAg** [comment]
Tests look great to me. Behavior is as expected. Let's land!
